fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,258 @@ impl ExecutorInner {
 
-        let mut transitions = Vec::new();
-        for (state, meta) in active_states {
-            match self
-                .get_transition_for(state, meta, global_context_gen)
-                .await
-            {
-                None => {}
-                Some(t) => transitions.push(t),
-            }
+        /// All futures in the executor resolve to this type, so the handling
+        /// code can tell them apart.
+        enum ExecutorLoopEvent {
+            /// Notivication about `DynState` arrived and should be handled,
+            /// usually added to the list of pending futures.
+            New {
+                state: DynState,
+                sm_update_rx: mpsc::UnboundedReceiver<DynState>,
+            },
+            /// One of trigger functions of a state machine finished and
+            /// returned transition function to run
+            Triggered(TransitionForActiveState),
+            /// Transition function and all the accounting around it are done
+            Completed {
+                state: DynState,
+                outcome: ActiveOrInactiveState,
+            },
+            /// New job receiver disconnected, that can only mean termination
+            Disconnected,
         }
 
-        loop {
-            if active_state_count == 0 {
-                debug!(
-                    "No state transitions remaining, exiting execute_next_state_transitions loops"
-                );
-                return Ok(());
-            }
-            let num_states = active_state_count;
-            let num_transitions = transitions.len();
-            debug!(
-                num_states,
-                num_transitions, "Awaiting any state transition to become ready"
-            );
-            let new_state_added = async move {
-                loop {
-                    // Prioritize existing active states over new states
-                    fedimint_core::task::sleep(EXECUTOR_POLL_INTERVAL).await;
-                    let new_active_states_count = self.get_active_states().await.len();
-                    if new_active_states_count > active_state_count {
-                        return;
+        // Keeps track of things already running, so we can deduplicate, just
+        // in case.
+        let mut currently_running_sms = HashSet::<DynState>::new();
+        // All things happening in parallel go into here
+        let mut futures: FuturesUnordered<BoxFuture<'_, ExecutorLoopEvent>> =
+            FuturesUnordered::new();
+
+        re_add_sm_update_receiver(&mut futures, sm_update_rx);
+
+        // main reactor loop: wait for next thing that completed, react (possibly adding
+        // more things to `futures`)
+        while let Some(event) = futures.next().await {
+            match event {
+                ExecutorLoopEvent::New {
+                    state,
+                    sm_update_rx,
+                } => {
+                    re_add_sm_update_receiver(&mut futures, sm_update_rx);
```

Comment:
Instead of running the `sm_update_rx.recv()` future inside `futures`, why not do a `select!` over both `futures.next()` and `sm_update_rx.recv()`? That way you wouldn't need to awkwardly re-add that receive future all the time.

Reply:
True. I wanted to keep everything going through these `FuturesUnurdered` that I missed that I can still compose it.